File: ros2/py_service/py_service/oop_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial


class MyNode(Node):
    def __init__(self):
        super().__init__("py_oop_client")
        self.get_logger().info("oop_client:")
        self.call_add_two_ints_server(12,20)
        self.call_add_two_ints_server(32,20)


    def call_add_two_ints_server(self, a, b):
        client_ = self.create_client(AddTwoInts, "add")
        # Service oluşana kadar bu kısımda bekler
        while not client_.wait_for_service(1.0):
            self.get_logger.warn("Waiting for Server ...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        # Veriyi gönderir
        future = client_.call_async(request)
        future.add_done_callback(partial(self.addCallback, a=a, b=b))

    def addCallback(self, future, a, b):

        try:
            response = future.result()
            self.get_logger().info(str(a)+"+"+str(b)+"="+str(response.sum))

        except Exception as e:
            self.get_logger.error("Server call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    rclpy.spin(MyNode())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2/cpp_service/src/oop_client.cpp
#include "my_oop_client.hpp"

Client::Client(): Node("oop_client_node"){
  client_ = this->create_client<add2Int>("server_test");
  threads.push_back(std::thread(std::bind(&Client::clientCallback, this, 10, 23)));
  threads.push_back(std::thread(std::bind(&Client::clientCallback, this, 15, 23)));
  threads.push_back(std::thread(std::bind(&Client::clientCallback, this, 1, 23)));
  RCLCPP_DEBUG(this->get_logger(),"OOP Client has been started.");
  threads.push_back(std::thread(std::bind(&Client::clientCallback, this, 10, 223)));
}

Client::~Client(){
  RCLCPP_DEBUG(this->get_logger(), "OOP Client terminated...");
}

void Client::clientCallback(int a, int b){
  while(!client_->wait_for_service(std::chrono::seconds(1))){
    RCLCPP_WARN(this->get_logger(), "Wait for the server to be up...");
  }
  auto request = std::make_shared<add2Int::Request>();
  request->a = a;
  request->b = b;
  auto future = client_->async_send_request(request);
  try
  {
      // auto response = future.get();
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Sum: %ld", future.get()->sum);
  }
  catch(const std::exception& e)
  {
      RCLCPP_ERROR(this->get_logger(), "Failed to call service add_two_ints");
  }
}

int main(int argc, char ** args){
    rclcpp::init(argc, args);
    rclcpp::spin(std::make_shared<Client>());
    rclcpp::shutdown();
    return 0;
}